### PR TITLE
[SPARK-5473] [EC2] Expose SSH failures after status checks pass

### PR DESCRIPTION
**Pull request by nchammas**
If there is some fatal problem with launching a cluster, `spark-ec2` just hangs without giving the user useful feedback on what the problem is.

This PR exposes the output of the SSH calls to the user if the SSH test fails during cluster launch for any reason but the instance status checks are all green. It also removes the growing trail of dots while waiting in favor of a fixed 3 dots.

For example:

```
$ ./ec2/spark-ec2 -k key -i /incorrect/path/identity.pem --instance-type m3.medium --slaves 1 --zone us-east-1c launch "spark-test"
Setting up security groups...
Searching for existing cluster spark-test...
Spark AMI: ami-35b1885c
Launching instances...
Launched 1 slaves in us-east-1c, regid = r-7dadd096
Launched master in us-east-1c, regid = r-fcadd017
Waiting for cluster to enter 'ssh-ready' state...
Warning: SSH connection error. (This could be temporary.)
Host: 127.0.0.1
SSH return code: 255
SSH output: Warning: Identity file /incorrect/path/identity.pem not accessible: No such file or directory.
Warning: Permanently added '127.0.0.1' (RSA) to the list of known hosts.
Permission denied (publickey).
```

This should give users enough information when some unrecoverable error occurs during launch so they can know to abort the launch. This will help avoid situations like the ones reported [here on Stack Overflow](http://stackoverflow.com/q/28002443/) and [here on the user list](http://mail-archives.apache.org/mod_mbox/spark-user/201501.mbox/%3C1422323829398-21381.post@n3.nabble.com%3E), where the users couldn't tell what the problem was because it was being hidden by `spark-ec2`.

This is a usability improvement that should be backported to 1.2.

Resolves [SPARK-5473](https://issues.apache.org/jira/browse/SPARK-5473).

### CODE REVIEW

**Comment by nchammas** [comment]
cc @shivaram @JoshRosen

**Comment by shivaram** [comment]
The idea of printing the error message sounds good to me. FWIW I actually found the dots useful to figure out how long the script was going to wait before trying again -- Will we just see an SSH error message now in place ?

**Comment by nchammas** [comment]
> FWIW I actually found the dots useful to figure out how long the script was going to wait before trying again

I find this useful too, but it doesn't work when we throw in the SSH output, since the dots and SSH output mix. I could look into a solution for that.

> Will we just see an SSH error message now in place ?

Now we'll just see 3 dots followed by the start of a cluster build if everything goes well, and 3 dots followed by SSH output if something is potentially broken (as in the example above). In the former case the wait time until `ssh-ready` will still be printed as before.

**Comment by nchammas** [comment]
I'm thinking it would be useful for a separate PR to add some information about how many nodes of the cluster are up during a cluster launch like this:

```
Waiting for cluster to enter 'ssh-ready' state... ( 1/30)
...
Waiting for cluster to enter 'ssh-ready' state... (29/30)
Waiting for cluster to enter 'ssh-ready' state... (30/30)
```

Or maybe just a simple spinner:

```
Waiting for cluster to enter 'ssh-ready' state... /
Waiting for cluster to enter 'ssh-ready' state... -
Waiting for cluster to enter 'ssh-ready' state... \
Waiting for cluster to enter 'ssh-ready' state... |
Waiting for cluster to enter 'ssh-ready' state... /
...
```

The original trail of dots also works.

In all cases we need to figure out how to do that neatly while allowing SSH output to also be printed to screen.

**Comment by nchammas** [comment]
@shivaram Does this solution work for you, or do you want the growing dots back?

@pwendell Does this meet your suggestion in [#4162](https://github.com/apache/spark/pull/4162#issuecomment-72113135)?

**Comment by shivaram** [comment]
I haven't tried out this solution, so I am not exactly sure what gets printed (I can do it over the weekend sometime). At a high-level my comment is that every attempt that checks if the cluster is `ssh-ready` should print some feedback on the screen so the user knows the script is not hun. If that is the case I'm fine with this solution.

**Comment by nchammas** [comment]
You can see example output in [the PR description](https://github.com/apache/spark/pull/4262#issue-55856344).

I will look into adding feedback while the script is waiting on the cluster to reach a certain state.

**Comment by nchammas** [comment]
OK, I added the dots back in. It doesn't look great, but it's serviceable:

```
$ ./ec2/spark-ec2 -k key -i /good/path/pem --instance-type m3.medium --slaves 40 --zone us-east-1c launch "spark-test"
Setting up security groups...
Searching for existing cluster spark-test...
Spark AMI: ami-35b1885c
Launching instances...
Launched 40 slaves in us-east-1c, regid = r-811e646a
Launched master in us-east-1c, regid = r-501c66bb
Waiting for cluster to enter 'ssh-ready' state...........

Warning: SSH connection error. (This could be temporary.)
Host: 54.54.54.54
SSH return code: 255
SSH output: ssh: connect to host 54.54.54.54 port 22: Connection refused

.

Warning: SSH connection error. (This could be temporary.)
Host: 54.54.54.54
SSH return code: 255
SSH output: ssh: connect to host 54.54.54.54 port 22: Connection refused

.
Cluster is now in 'ssh-ready' state. Waited 517 seconds.
Generating cluster's SSH key on master...
```

That's what we get in the case of a temporary failure while SSH is still coming up.

In the case of a permanent failure, you just get the repeated `.` on a separate line followed by the 4-line error report:

```
$ ./ec2/spark-ec2 -k key -i /wrong/path/pem --instance-type m3.medium --slaves 1 --zone us-east-1c launch "spark-test"
Setting up security groups...
Searching for existing cluster spark-test...
Spark AMI: ami-35b1885c
Launching instances...
Launched 1 slaves in us-east-1c, regid = r-283248c3
Launched master in us-east-1c, regid = r-2b3248c0
Waiting for cluster to enter 'ssh-ready' state..............

Warning: SSH connection error. (This could be temporary.)
Host: 54.54.54.54
SSH return code: 255
SSH output: Warning: Identity file /wrong/path/pem not accessible: No such file or directory.
Warning: Permanently added '54.54.54.54' (RSA) to the list of known hosts.
Permission denied (publickey).

.

Warning: SSH connection error. (This could be temporary.)
Host: 54.54.54.54
SSH return code: 255
SSH output: Warning: Identity file /wrong/path/pem not accessible: No such file or directory.
Permission denied (publickey).

.

...
```

What do you think?

File: ec2/spark_ec2.py [hunk 1]
```diff
@@ -675,29 +676,40 @@ def setup_spark_cluster(master, opts):
...
+        """).format(
+            h=host,
+            r=s.returncode,
+            o=cmd_output.strip()
+        )
+
+    return s.returncode == 0
+
+
+def is_cluster_ssh_available(cluster_instances, opts, print_ssh_output=False):
```
Comment (nchammas) [inline ec2/spark_ec2.py]: I think I'll change this to default to `print_ssh_output=True`.

**Comment by nchammas** [comment]
This PR is ready for a final review.

**Comment by srowen** [comment]
@shivaram Do you have any more comments on this change?

**Comment by shivaram** [comment]
I just tried out the latest version and it worked fine - LGTM. Thanks @nchammas for the change.

**Pull request closed without merging**